src/compiler/crystal/syntax/parser.cr

```diff
@@ -2403,19 +2406,24 @@ module Crystal
 
                   tuple = TupleLiteral.new(tuple_elements).at(curly_location)
                   when_conds << tuple
+                  add_when_exp(when_exps, tuple)
 
                   check :"}"
                   next_token_skip_space
                 else
-                  when_conds << parse_when_expression(cond)
+                  exp = parse_when_expression(cond)
+                  when_conds << exp
+                  add_when_exp(when_exps, exp)
                   skip_space
                 end
 
                 break if when_expression_end
               end
             else
               while true
-                when_conds << parse_when_expression(cond)
+                exp = parse_when_expression(cond)
+                when_conds << exp
+                add_when_exp(when_exps, exp)
```

Comment:
`when_conds << exp` shouldn't be done here, as it would cause a raise in `add_when_exp`. `if when_exps.includes?(exp)` will be true always.

Reply:
`add_when_exp` tests against `when_exps`, not against `when_conds`, I don't know if that's what you mean

Reply:
Oh, I misread it!! My mistake